### PR TITLE
CLOUDSTACK-4045 IP address acquired with associateIpAddress is marked as source NAT

### PR DESCRIPTION
**Pull request by houthuis**
added a check for network state when determining whether a new IP should be source NAT. this prevents associated IP's to be marked as source NAT when the network is in allocated state, causing disassociateIpAddress to fail later

Code will now throw a InvalidParameterValueException in the above scenario.

### CODE REVIEW

**Comment by PaulAngus** [comment]
@blueorangutan package

File: server/src/com/cloud/network/IpAddressManagerImpl.java [hunk 1]
```diff
@@ -1394,6 +1380,25 @@ public IPAddressVO associateIPToGuestNetwork(long ipId, long networkId, boolean
         }
     }
 
+    protected boolean isSourceNatAvailableForNetwork(long networkId, Account owner, IPAddressVO ipToAssoc, Network network) {
```
Comment (DaanHoogland) [inline server/src/com/cloud/network/IpAddressManagerImpl.java]: is networkId needed on this signature? network is being passed as well!

File: server/test/com/cloud/network/IpAddressManagerTest.java [hunk 2]
```diff
@@ -17,38 +17,109 @@
...
+
+    NetworkVO networkNat;
+
+    AccountVO account;
+
     @Before
-    public void setup() {
+    public void setup() throws ResourceUnavailableException {
         MockitoAnnotations.initMocks(this);
+
```
Comment (DaanHoogland) [inline server/test/com/cloud/network/IpAddressManagerTest.java]: a lot of this mock data is test specific and should not be in the setup method

File: server/test/com/cloud/network/IpAddressManagerTest.java [hunk 3]
```diff
@@ -17,38 +17,109 @@
...
+        when(networkNat.getId()).thenReturn(3L);
+        when(networkNat.getVpcId()).thenReturn(null);
+
+        account = new AccountVO("admin", 1L, null, (short) 1, 1L, "c65a73d5-ebbd-11e7-8f45-107b44277808");
+        account.setId(1L);
+
+        NetworkOfferingVO networkOfferingVO = Mockito.mock(NetworkOfferingVO.class);
+        networkOfferingVO.setSharedSourceNat(false);
+
+        Mockito.when(_networkDao.findById(1L)).thenReturn(networkAllocated);
```
Comment (DaanHoogland) [inline server/test/com/cloud/network/IpAddressManagerTest.java]: the ids of networkAllocated and networkImplemented are mixed up in comparison with the test methods

File: server/src/com/cloud/network/IpAddressManagerImpl.java (hunk 1 above)
Comment (houthuis) [reply server/src/com/cloud/network/IpAddressManagerImpl.java]: no, not really. must be a result of intellij's extract method

**Comment by houthuis** [comment]
@blueorangutan package

**Comment by houthuis** [comment]
@blueorangutan test

**Comment by rhtyd** [comment]
@houthuis looks like we're at capacity, we currently don't have you in the `bo` ACLs to allow kicking tests.
Use @blueorangutan help to see options and ACl list.

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Comment by rhtyd** [comment]
@houthuis looks like some failures both in travis and trillian results, can you check?

**Comment by houthuis** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Comment by DaanHoogland** [comment]
looks like the tests are ok, as well @houthuis . let's rerun after the stabalisation fixes and last features for 4.11 are merged. but given no conflicts this can go in right after 4.11 is branched off. Pleae edit the title btw. Just a issue number is a bit meagre. something like "acquire associateIpAddress as SNAT" on the side? Maybe you/we can do better.

**Comment by borisstoyanov** [comment]
@houthuis can you please resolve the merge conflicts and we can run tests on it again

**Comment by DaanHoogland** [comment]
@blueorangutan package

File: server/src/main/java/com/cloud/network/IpAddressManagerImpl.java [hunk 4]
```diff
@@ -1417,6 +1408,25 @@ public IPAddressVO associateIPToGuestNetwork(long ipId, long networkId, boolean
         }
     }
 
+    protected boolean isSourceNatAvailableForNetwork(Account owner, IPAddressVO ipToAssoc, Network network) {
```
Comment (rafaelweingartner) [inline server/src/main/java/com/cloud/network/IpAddressManagerImpl.java]: To make it perfect a Java documentation describing when an IP is considered source NAT would be welcome.

**Comment by rafaelweingartner** [comment]
@blueorangutan package.

**Comment by DaanHoogland** [comment]
@houthuis can you considder @rafaelweingartner 's comment?
@blueorangutan test

**Comment by houthuis** [comment]
@DaanHoogland thanks for running the tests, I am happy.

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan test

**Comment by DaanHoogland** [comment]
I think it is a bugfix, but no biggy. tried to just retarget but that pulls in unwanted stuff: rebase needed! @houthuis @rhtyd

**Comment by rhtyd** [comment]
Ping @houthuis

**Comment by rhtyd** [comment]
can you fix the conflict @houthuis

**Comment by houthuis** [comment]
> can you fix the conflict @houthuis

@rohit done, awaiting test results

**Comment by DaanHoogland** [comment]
@houthuis can you rebase again?

**Comment by GabrielBrascher** [comment]
@houthuis can you please rebase against the latest master?

**Comment by rafaelweingartner** [comment]
@blueorangutan package

**Comment by rafaelweingartner** [comment]
@blueorangutan package

**Comment by GabrielBrascher** [comment]
@houthuis there are some errors on the build. Could you please take a look and fix? Thanks!

```
[ERROR]   IpAddressManagerTest.isIpEqualsGatewayOrNetworkOfferingsEmptyTestExpectFalseServicesCidrNotNull:153 
Wanted but not invoked:
networkModel.listNetworkOfferingServices(
    <any>
);
-> at com.cloud.network.IpAddressManagerTest.isIpEqualsGatewayOrNetworkOfferingsEmptyTestExpectFalseServicesCidrNotNull(IpAddressManagerTest.java:153)
Actually, there were zero interactions with this mock.

[ERROR]   IpAddressManagerTest.isIpEqualsGatewayOrNetworkOfferingsEmptyTestExpectFalseServicesNotEmpty:143 
Wanted but not invoked:
networkModel.listNetworkOfferingServices(
    <any>
);
-> at com.cloud.network.IpAddressManagerTest.isIpEqualsGatewayOrNetworkOfferingsEmptyTestExpectFalseServicesNotEmpty(IpAddressManagerTest.java:143)
Actually, there were zero interactions with this mock.

[ERROR]   IpAddressManagerTest.isIpEqualsGatewayOrNetworkOfferingsEmptyTestNetworkOfferingsEmptyAndCidrNull:221 
Wanted but not invoked:
networkModel.listNetworkOfferingServices(
    <any>
);
-> at com.cloud.network.IpAddressManagerTest.isIpEqualsGatewayOrNetworkOfferingsEmptyTestNetworkOfferingsEmptyAndCidrNull(IpAddressManagerTest.java:221)
Actually, there were zero interactions with this mock.
```

**Comment by DaanHoogland** [comment]
closing as work is continued in #3125

File: server/src/main/java/com/cloud/network/IpAddressManagerImpl.java (hunk 4 above)
Comment (rafaelweingartner) [reply server/src/main/java/com/cloud/network/IpAddressManagerImpl.java]: Well, for me, the java doc should present  the idea of the method, and not just describe the code. The code is the implementation of the idea. By idea here I mean, the context where the given method fits in, the behavior we intend it to have (from an input/output perspective), but not describing the implementation at a line by line perspective.

**Pull request closed without merging**